tutorial/function-pointers-as-arguments.md

```diff
@@ -0,0 +1,441 @@
+  implicit none
+
+  contains
+
+  !........................................
+  function high_int_int_1(function1, function2, a) result(Out_0001)
+
+    implicit none
+
+    integer(i64) :: Out_0001
+    integer(i64), value :: a
+    integer(i64) :: x
+    integer(i64) :: y
+
+    interface
+      function function1(in_0000) result(out_0000)
+        use, intrinsic :: ISO_C_Binding, only : i64 => C_INT64_T
+        integer(i64) :: out_0000
+        integer(i64), value :: in_0000
+      end function function1
+
+      function function2(in_0001) result(out_0001)
+        use, intrinsic :: ISO_C_Binding, only : i64 => C_INT64_T
+        integer(i64) :: out_0001
+        integer(i64), value :: in_0001
+      end function function2
+    end interface
+
+    x = function1(a)
+    y = function2(a)
+    Out_0001 = x + y
+    return
+
+  end function high_int_int_1
+  !........................................
+
+end module boo
+```
+
+Now, we will see a special case that is optimized by Pyccel (not supported in C yet):
+
+In this the python code Pyccel will recognize that `foo` doesn't change `x`, so it will automatically add `const` or `intent` (depanding on the language C/Fortran) to the data type of `x`
+providing a useful information for C/Fortran compilers to make some optimizations on the code. Also:
+
+```python
+def foo(x: 'int[:]', i: 'int'):
+    #some code
+    return x[i]
+
+def func1(a: 'int[:]', b: 'int', func_arg : '(int)(int[:], int)' = foo):
+    x = foo(a, b)
+    return x
+```
+
+The Fortran equivalent:
+
+```Fortran
+module boo
+
+
+  use, intrinsic :: ISO_C_Binding, only : i64 => C_INT64_T
+  implicit none
+
+  contains
+
+  !........................................
+  function foo(x, i) result(Out_0001)
+
+    implicit none
+
+    integer(i64) :: Out_0001
+    integer(i64), intent(in) :: x(0:)
+    integer(i64), value :: i
+
+    !some code
+    Out_0001 = x(i)
+    return
+
+  end function foo
+  !........................................
+
+  !........................................
+  function func1(a, b, func_arg) result(x)
+
+    implicit none
+
+    integer(i64) :: x
+    integer(i64), intent(in) :: a(0:)
+    integer(i64), value :: b
+
+    interface
+      function func_arg(in_0000, in_0001) result(out_0000)
+        use, intrinsic :: ISO_C_Binding, only : i64 => C_INT64_T
+        integer(i64) :: out_0000
+        integer(i64), intent(inout) :: in_0000(0:)
+        integer(i64), value :: in_0001
+      end function func_arg
+    end interface
+
+    x = foo(a, b)
+    return
+
+  end function func1
+  !........................................
+
+end module boo
+```
+
+Note that the argument in the interface in `func1` has a differente [intent](https://pages.mtu.edu/~shene/COURSES/cs201/NOTES/chap07/intent.html) the argument `x` in `foo` that shouldn't be `intent(in)`, But as Pyccel detected that `x` wont change in `foo` the perfect case for `x` is to be an `intent(in)` rather than `intent(inout)`.
```

Comment:
Spelling/Grammar:
> Note that the argument in the interface in `func1` has a different [intent](https://pages.mtu.edu/~shene/COURSES/cs201/NOTES/chap07/intent.html). The argument `x` in `foo` shouldn't be `intent(in)`, but rather `intent(inout)`. However as Pyccel detected that `x` wont change in `foo`, the perfect case for `x` is to be an `intent(in)` rather than `intent(inout)`.